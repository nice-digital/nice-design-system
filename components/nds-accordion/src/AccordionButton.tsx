import React, { FC, ReactNode, SyntheticEvent } from "react";
import { Toggle } from "./Toggle";
import WarningIcon from "@nice-digital/icons/lib/Warning";
import { accordionVariants } from "./Accordion";

export interface AccordionButtonProps {
	isOpen: boolean;
	toggleAccordion: (e: SyntheticEvent<HTMLButtonElement>) => void;
	generatedId: string;
	showLabel: string;
	hideLabel: ReactNode;
	title: ReactNode;
	variant?: keyof typeof accordionVariants;
}

export const AccordionButton: FC<AccordionButtonProps> = ({
	isOpen,
	toggleAccordion,
	generatedId,
	showLabel,
	hideLabel,
	title,
	variant = "default"
}) => {
	return (
		<button
			aria-expanded={isOpen}
			aria-controls={generatedId}
			onClick={toggleAccordion}
			data-tracking={isOpen ? hideLabel : showLabel}
			className="accordion__summary"
			type="button"
		>
			<Toggle isOpen={isOpen} className={"accordion__toggleLabel"}>
				{isOpen ? hideLabel : showLabel}
			</Toggle>
			<span className="accordion__title">
				{variant === "caution" && <WarningIcon className="warning-icon" />}
				{title}
			</span>
		</button>
	);
};
