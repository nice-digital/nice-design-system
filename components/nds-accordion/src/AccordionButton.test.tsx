import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { AccordionButton } from "./AccordionButton";

describe("AccordionButton", () => {
	const mockToggleAccordion = jest.fn();
	const mockGeneratedId = "accordion-id";
	const mockShowLabel = "Show";
	const mockHideLabel = "Hide";
	const mockTitle = "Test Title";
	const mockVariant = "caution";

	it("should render correctly when accordion is open", () => {
		render(
			<AccordionButton
				isOpen={true}
				toggleAccordion={mockToggleAccordion}
				generatedId={mockGeneratedId}
				showLabel={mockShowLabel}
				hideLabel={mockHideLabel}
				title={mockTitle}
			/>
		);

		expect(screen.getByRole("button")).toHaveAttribute("aria-expanded", "true");
		expect(screen.getByRole("button")).toHaveAttribute(
			"aria-controls",
			mockGeneratedId
		);
		expect(screen.getByRole("button")).toMatchSnapshot();
	});

	it("should render correctly when accordion is closed", () => {
		render(
			<AccordionButton
				isOpen={false}
				toggleAccordion={mockToggleAccordion}
				generatedId={mockGeneratedId}
				showLabel={mockShowLabel}
				hideLabel={mockHideLabel}
				title={mockTitle}
			/>
		);
		expect(screen.getByRole("button")).toHaveAttribute(
			"aria-expanded",
			"false"
		);
		expect(screen.getByRole("button")).toHaveAttribute(
			"aria-controls",
			mockGeneratedId
		);
		expect(screen.getByRole("button")).toMatchSnapshot();
	});

	it("should call toggleAccordion when clicked", () => {
		render(
			<AccordionButton
				isOpen={true}
				toggleAccordion={mockToggleAccordion}
				generatedId={mockGeneratedId}
				showLabel={mockShowLabel}
				hideLabel={mockHideLabel}
				title={mockTitle}
			/>
		);

		fireEvent.click(screen.getByRole("button"));
		expect(mockToggleAccordion).toHaveBeenCalledTimes(1);
	});

	it("should render correctly when variant is caution", () => {
		render(
			<AccordionButton
				isOpen={true}
				toggleAccordion={mockToggleAccordion}
				generatedId={mockGeneratedId}
				showLabel={mockShowLabel}
				hideLabel={mockHideLabel}
				title={mockTitle}
				variant={mockVariant}
			/>
		);

		expect(screen.getByRole("button")).toMatchSnapshot();
	});
});
