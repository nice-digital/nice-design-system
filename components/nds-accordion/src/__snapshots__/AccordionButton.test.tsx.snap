// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionButton should render correctly when accordion is closed 1`] = `
<button
  aria-controls="accordion-id"
  aria-expanded="false"
  class="accordion__summary"
  data-tracking="Show"
  type="button"
>
  <span
    class="accordion__toggleLabel toggle__label"
  >
    <svg
      aria-hidden="true"
      class="toggle__icon"
      fill="none"
      focusable="false"
      height="1em"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
        fill="currentColor"
      />
    </svg>
    Show
  </span>
   
  <span
    class="accordion__title"
  >
    Test Title
  </span>
</button>
`;

exports[`AccordionButton should render correctly when accordion is open 1`] = `
<button
  aria-controls="accordion-id"
  aria-expanded="true"
  class="accordion__summary"
  data-tracking="Hide"
  type="button"
>
  <span
    class="accordion__toggleLabel toggle__label"
  >
    <svg
      aria-hidden="true"
      class="toggle__icon toggle__icon--open"
      fill="none"
      focusable="false"
      height="1em"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
        fill="currentColor"
      />
    </svg>
    Hide
  </span>
   
  <span
    class="accordion__title"
  >
    Test Title
  </span>
</button>
`;

exports[`AccordionButton should render correctly when variant is caution 1`] = `
<button
  aria-controls="accordion-id"
  aria-expanded="true"
  class="accordion__summary"
  data-tracking="Hide"
  type="button"
>
  <span
    class="accordion__toggleLabel toggle__label"
  >
    <svg
      aria-hidden="true"
      class="toggle__icon toggle__icon--open"
      fill="none"
      focusable="false"
      height="1em"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
        fill="currentColor"
      />
    </svg>
    Hide
  </span>
   
  <span
    class="accordion__title"
  >
    <svg
      aria-hidden="true"
      class="warning-icon"
      fill="none"
      focusable="false"
      height="1em"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M429.067 456.4H82.342c-20.328 0-36.703-8.151-44.91-22.369-8.195-14.205-7.081-32.449 3.097-50.05L213.89 83.695C224.07 66.093 239.302 56 255.718 56c16.417 0 31.649 10.107 41.814 27.709L470.88 383.995c10.178 17.602 11.306 35.845 3.097 50.05-8.208 14.204-24.568 22.355-44.91 22.355ZM255.718 92.445c-2.384 0-6.467 2.927-10.249 9.493L72.106 402.21c-3.783 6.553-4.282 11.563-3.097 13.619 1.185 2.056 5.767 4.126 13.347 4.126h346.739c7.566 0 12.148-2.084 13.348-4.14 1.199-2.056.685-7.052-3.098-13.619L265.983 101.91c-3.783-6.552-7.88-9.479-10.25-9.479l-.015.014Zm-21.641 202.898c.956 9.95 2.57 17.274 4.939 22.385 2.97 6.438 8.48 9.835 15.946 9.835 7.309 0 12.848-3.44 16.046-9.964 2.598-5.31 4.24-12.534 4.982-22.013l6.424-73.633a200.589 200.589 0 0 0 1.071-20.571c0-12.078-1.585-21.2-4.854-27.923-2.627-5.396-8.566-11.821-21.87-11.821-8.551 0-15.589 2.898-20.928 8.594-5.254 5.611-7.909 13.32-7.909 22.927 0 6.195.457 16.417 1.356 30.407l4.797 71.777Zm1.884 56.004c5.325-5.325 11.906-8.038 19.601-8.038 7.651 0 14.275 2.684 19.7 7.98 5.467 5.339 8.223 11.906 8.223 19.501 0 8.522-2.856 15.36-8.466 20.357-5.453 4.854-11.891 7.295-19.157 7.295-7.352 0-13.862-2.413-19.344-7.181-5.696-4.94-8.594-11.835-8.594-20.457 0-7.58 2.712-14.133 8.037-19.457Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    Test Title
  </span>
</button>
`;
